From 9e409d567fbb2ebc617d29fe544346ddd40c97f7 Mon Sep 17 00:00:00 2001
From: Igal Zeltser <izeltser@habana.ai>
Date: Thu, 01 Apr 2021 19:01:53 +0300
Subject: [PATCH] [SW-28013] module: pci: adding iatu initialization

Before, the pci address translation unit was configured by LKD,
from now on the iATU initialization is carried out by FW.

Testing showed that for some devices (ie Gaudi) the iATU must be
initialized after the pci memory access is enabled.
The pci module is monitoring the pci memory access bit, and when it's
enabled the regions of iatu are configured.

A flag bit is set in the status register to signal the LKD driver
that the fw is taking care of the iATU initialization.
This is done in an early stage (ppboot) so it would be set before
the LKD driver is reading it.

Change-Id: I690c42fe16933a7a4d2fbd626924d4bb7a64f913
Signed-off-by: Igal Zeltser <izeltser@habana.ai>
---

diff --git a/gaudi/asic.c b/gaudi/asic.c
index 78ccc3a..85638f1 100644
--- a/gaudi/asic.c
+++ b/gaudi/asic.c
@@ -26,6 +26,7 @@
 #include <boot_if.h>
 #include <hl_fw_if.h>
 #include <crc32.h>
+#include <pci.h>
 
 #include "common.h"
 
@@ -320,3 +321,32 @@
 
 	return false; /* No end */
 }
+
+void hl_pci_iatu_init(void)
+{
+	struct hl_pci_inbound_region inbound_region;
+	struct hl_pci_outbound_region outbound_region;
+
+	/* Inbound Region 0 - Bar 0 - Point to SRAM + CFG */
+	inbound_region.mode = PCI_BAR_MATCH_MODE;
+	inbound_region.bar = PCI_BAR_ID_0;
+	inbound_region.addr = SDRAM_BASE;
+	hl_pci_set_inbound_region(0, &inbound_region);
+
+	/* Inbound Region 1 - Bar 2 - Point to SPI FLASH */
+	inbound_region.mode = PCI_BAR_MATCH_MODE;
+	inbound_region.bar = PCI_BAR_ID_2;
+	inbound_region.addr = HL_SPI_BASE; //cfg_base
+	hl_pci_set_inbound_region(1, &inbound_region);
+
+	/* Inbound Region 2 - Bar 4 - Point to HBM */
+	inbound_region.mode = PCI_BAR_MATCH_MODE;
+	inbound_region.bar = PCI_BAR_ID_4;
+	inbound_region.addr = DRAM_BASE;
+	hl_pci_set_inbound_region(2, &inbound_region);
+
+	/* Outbound Region 0 - Point to Host */
+	outbound_region.addr = HOST_PHYS_BASE;
+	outbound_region.size = HOST_PHYS_SIZE;
+	hl_pci_set_outbound_region(&outbound_region);
+}
diff --git a/gaudi/common.c b/gaudi/common.c
index 56063af..9330ec2 100644
--- a/gaudi/common.c
+++ b/gaudi/common.c
@@ -304,7 +304,10 @@
 	}
 }
 
-void hl_plat_early_init(void) {}
+void hl_plat_early_init(void)
+{
+	cpu_boot_dev_sts_clrset(CPU_BOOT_DEV_STS0_FW_IATU_CONF_EN, true);
+}
 
 
 void ppboot_log_waiting_for_pcie(void)
diff --git a/gaudi/common.h b/gaudi/common.h
index e0cf8f9..adbef9a 100644
--- a/gaudi/common.h
+++ b/gaudi/common.h
@@ -22,11 +22,15 @@
 #define PPBOOT_TEXT_BASE		0x7FF0000000
 #define PREBOOT_TEXT_BASE		(PPBOOT_TEXT_BASE + PPBOOT_SIZE)
 
+#define DRAM_BASE			0x0
 #define HL_SPI_BASE			0x7FF8000000ull
 #define HL_CFG_BASE			0x7FFC000000ull
 #define SDRAM_BASE			0x7FF0000000ull
 #define FIT_IMG_BASE			0x7FF0200000ull
 
+#define HOST_PHYS_BASE			0x8000000000ull
+#define HOST_PHYS_SIZE			0x1000000000000ull
+
 #define FUSE_WORDS_PER_BANK		64
 #define I2C_SLAVE_ADDRESS		0x4f
 #define I2C_SLAVE_ADDRESS_DBG		0x47
diff --git a/gaudi2_arm/asic.c b/gaudi2_arm/asic.c
index a7e7715..c495d55 100644
--- a/gaudi2_arm/asic.c
+++ b/gaudi2_arm/asic.c
@@ -15,6 +15,7 @@
 #include <uart.h>
 #include <sha256.h>
 #include <boot_desc.h>
+#include <pci.h>
 
 #include "common.h"
 
@@ -75,3 +76,38 @@
 
 	return !memcmp(hash, desc->chksum, ARRAY_SIZE(hash));
 }
+
+void hl_pci_iatu_init(void)
+{
+	struct hl_pci_inbound_region inbound_region;
+	struct hl_pci_outbound_region outbound_region;
+
+	hl_pci_get_bar_phys_addr(PCI_BAR_ID_0, &inbound_region.pcie_bar_phys);
+
+	/* Inbound Region 0 - Bar 0 - Point to CFG */
+	inbound_region.mode = PCI_ADDRESS_MATCH_MODE;
+	inbound_region.bar = PCI_BAR_ID_0;
+	inbound_region.offset_in_bar = 0;
+	inbound_region.addr = STM_FLASH_BASE;
+	inbound_region.size = CFG_REGION_SIZE;
+	hl_pci_set_inbound_region(0, &inbound_region);
+
+	/* Inbound Region 1 - Bar 0 - Point to BAR0_RESERVED + SRAM */
+	inbound_region.mode = PCI_ADDRESS_MATCH_MODE;
+	inbound_region.bar = PCI_BAR_ID_0;
+	inbound_region.offset_in_bar = CFG_REGION_SIZE;
+	inbound_region.addr = BAR0_RSRVD_BASE;
+	inbound_region.size = BAR0_RSRVD_SIZE + SRAM_SIZE;
+	hl_pci_set_inbound_region(1, &inbound_region);
+
+	/* Inbound Region 2 - Bar 4 - Point to DRAM */
+	inbound_region.mode = PCI_BAR_MATCH_MODE;
+	inbound_region.bar = PCI_BAR_ID_4;
+	inbound_region.addr = DRAM_BASE;
+	hl_pci_set_inbound_region(2, &inbound_region);
+
+	/* Outbound Region 0 - Point to Host */
+	outbound_region.addr = HOST_PHYS_BASE_0;
+	outbound_region.size = HOST_PHYS_SIZE_0;
+	hl_pci_set_outbound_region(&outbound_region);
+}
diff --git a/gaudi2_arm/common.c b/gaudi2_arm/common.c
index efb2dab..469198d 100644
--- a/gaudi2_arm/common.c
+++ b/gaudi2_arm/common.c
@@ -34,6 +34,7 @@
 #include <string.h>
 #include <spi.h>
 #include <preboot/fit.h>
+#include <boot_if.h>
 
 #include "common.h"
 
@@ -404,6 +405,7 @@
 void hl_plat_early_init(void)
 {
 	init_arc();
+	cpu_boot_dev_sts_clrset(CPU_BOOT_DEV_STS0_FW_IATU_CONF_EN, true);
 }
 
 void ppboot_log_waiting_for_pcie(void)
diff --git a/gaudi2_arm/common.h b/gaudi2_arm/common.h
index 1243ae8..aba8975 100644
--- a/gaudi2_arm/common.h
+++ b/gaudi2_arm/common.h
@@ -19,10 +19,19 @@
 #define PPBOOT_TEXT_BASE		0x7FFD000000
 #define PREBOOT_TEXT_BASE		(PPBOOT_TEXT_BASE + PPBOOT_SIZE)
 
+#define CFG_REGION_SIZE			0xC000000
+#define BAR0_RSRVD_SIZE			0x1000000
+
 #define HL_SPI_BASE			0x3FF6000000ull
 #define HL_CFG_BASE			0x3FF8000000ull
 #define SDRAM_BASE			0x7FFD000000ull
 #define FIT_IMG_BASE			0x7FFD200000ull
+#define STM_FLASH_BASE			0x1000007FF4000000ull
+#define BAR0_RSRVD_BASE			0x1000FFFFFC000000ull
+#define DRAM_BASE			0x1001000000000000ull
+
+#define HOST_PHYS_BASE_0		0x0000000000000000ull
+#define HOST_PHYS_SIZE_0		0x0100000000000000ull
 
 #define HL_SP_SRAM_BASE			0x7FF7FE0000ULL
 #define HL_SP_SRAM_SIZE			0x10000
diff --git a/greco/asic.c b/greco/asic.c
index 0bd1d21..dd2eef7 100644
--- a/greco/asic.c
+++ b/greco/asic.c
@@ -15,6 +15,7 @@
 #include <uart.h>
 #include <sha256.h>
 #include <boot_desc.h>
+#include <pci.h>
 
 #include "common.h"
 
@@ -76,3 +77,25 @@
 	return !memcmp(hash, desc->chksum, ARRAY_SIZE(hash));
 }
 
+void hl_pci_iatu_init(void)
+{
+	struct hl_pci_inbound_region inbound_region;
+	struct hl_pci_outbound_region outbound_region;
+
+	/* Inbound Region 0 - Bar - - Point to SRAM and CFG */
+	inbound_region.mode = PCI_BAR_MATCH_MODE;
+	inbound_region.bar = PCI_BAR_ID_0;
+	inbound_region.addr = SDRAM_BASE;
+	hl_pci_set_inbound_region(0, &inbound_region);
+
+	/* Inbound Region 1 - Bar 4 - Point to DRAM */
+	inbound_region.mode = PCI_BAR_MATCH_MODE;
+	inbound_region.bar = PCI_BAR_ID_4;
+	inbound_region.addr = DRAM_BASE;
+	hl_pci_set_inbound_region(1, &inbound_region);
+
+	/* Outbound Region 0 - Point to Host */
+	outbound_region.addr = HOST_PHYS_BASE;
+	outbound_region.size = HOST_PHYS_SIZE;
+	hl_pci_set_outbound_region(&outbound_region);
+}
diff --git a/greco/common.c b/greco/common.c
index 3774df8..a53324d 100644
--- a/greco/common.c
+++ b/greco/common.c
@@ -39,6 +39,7 @@
 #include <spi.h>
 #include <preboot/fit.h>
 #include <efuse.h>
+#include <boot_if.h>
 
 #include "common.h"
 
@@ -494,6 +495,7 @@
 void hl_plat_early_init(void)
 {
 	axi_enable_drain();
+	cpu_boot_dev_sts_clrset(CPU_BOOT_DEV_STS0_FW_IATU_CONF_EN, true);
 }
 
 void ppboot_log_waiting_for_pcie(void)
diff --git a/greco/common.h b/greco/common.h
index 9e657e0..467cabc 100644
--- a/greco/common.h
+++ b/greco/common.h
@@ -33,6 +33,10 @@
 #define HL_CFG_BASE			0x7FFC000000ull
 #define SDRAM_BASE			0x7FF0000000ull
 #define FIT_IMG_BASE			0x7FF0200000ull
+#define DRAM_BASE			0x1000000000000ull
+
+#define HOST_PHYS_BASE			0x2000000000000ull
+#define HOST_PHYS_SIZE			0x2000000000000ull
 
 #define FUSE_WORDS_PER_BANK		0x40
 #define I2C_SLAVE_ADDRESS		0x6A
diff --git a/include/preboot.h b/include/preboot.h
index 8b38daf..5e7c464 100644
--- a/include/preboot.h
+++ b/include/preboot.h
@@ -151,6 +151,7 @@
 void preboot_disable_iatu_access(void);
 const char *preboot_get_version(enum preboot_vers ver);
 void hl_ram_init(void);
+void hl_pci_iatu_init(void);
 bool hl_pci_link_is_up(void);
 
 void preboot_finalize_fw_info(void);
diff --git a/preboot/module_pci.c b/preboot/module_pci.c
index 3d67caa..1bba904 100644
--- a/preboot/module_pci.c
+++ b/preboot/module_pci.c
@@ -12,6 +12,7 @@
 
 #include "cmd_module.h"
 #include "uart.h"
+#include "pci.h"
 
 struct pci_status {
 	u32 flr_reg_val;
@@ -19,6 +20,7 @@
 	u32 smlh_link_reg_val;
 	u32 rdlh_link_reg_val;
 	bool link_is_up;
+	bool mem_enabled;
 };
 
 static struct pci_status pci_sts = {
@@ -27,7 +29,7 @@
 	.smlh_link_reg_val = 1,
 	.rdlh_link_reg_val = 1,
 	.link_is_up = true,
-
+	.mem_enabled = false
 };
 
 __weak void preboot_pcie_exec_hw_wa(void)
@@ -46,6 +48,8 @@
 	regmap_writel(0x10, mmPCIE_CORE_HOT_RST);
 	regmap_writel(1, mmPCIE_AUX_LTSSM_EN);
 }
+
+__weak void hl_pci_iatu_init(void) {}
 
 void module_pci_setup(void)
 {
@@ -67,6 +71,9 @@
 	    pci_sts.link_is_up)
 		return FOUND_FAST_RETURN;
 
+	if (hl_pci_is_mem_enabled() != pci_sts.mem_enabled)
+		return FOUND_FAST_RETURN;
+
 	return NOT_FOUND;
 }
 
@@ -86,6 +93,8 @@
 
 void module_pci_run(void)
 {
+	bool pci_mem_enabled;
+
 	if (pci_sts.flr_reg_val == 1) {
 		regmap_writel(1, mmPCIE_AUX_PF_FLR_ACTIVE);
 		regmap_writel(1, mmPCIE_AUX_FLR_INT);
@@ -109,6 +118,13 @@
 			preboot_puts("down\n");
 		}
 	}
+
+	pci_mem_enabled = hl_pci_is_mem_enabled();
+	if (pci_mem_enabled != pci_sts.mem_enabled) {
+		pci_sts.mem_enabled = pci_mem_enabled;
+		if (pci_mem_enabled)
+			hl_pci_iatu_init();
+	}
 }
 
 void module_pci_enable(void)
